prow/spyglass/lenses/buildlog/lens.go

```diff
@@ -245,17 +277,107 @@ func (lens Lens) Callback(artifacts []api.Artifact, resourceDir string, data str
 	if !ok {
 		return fmt.Sprintf(missingArtifact, request.Artifact)
 	}
+	if request.Analyze {
+		conf := getConfig(rawConfig)
+		hr, err := analyzeArtifact(artifact, &conf)
+		if err != nil {
+			hr = &highlightResponse{Error: err.Error()}
+		}
+		buf, err := json.Marshal(hr)
+		if err != nil {
+			return err.Error()
+		}
+		return string(buf)
+	}
 	if request.SaveEnd != nil {
 		return storeHighlightedLines(&request, artifact)
 	}
 	return loadLines(&request, artifact, resourceDir, rawConfig)
 }
 
-func storeHighlightedLines(request *lineRequest, artifact api.Artifact) string {
-	err := artifact.UpdateMetadata(map[string]string{
-		focusStart: strconv.Itoa(request.StartLine),
-		focusEnd:   strconv.Itoa(*request.SaveEnd),
+type highlightRequest struct {
+	// URL to highlight
+	URL string `json:"url"`
+	// Pin if the highlight should be saved
+	Pin bool `json:"pin"`
+	// Overwrite if an existing highlight should be replaced
+	Overwrite bool `json:"overwrite"`
+}
+
+type highlightResponse struct {
+	// Min line number to highlight
+	Min int `json:"min"`
+	// Max line number to highlight (inclusive).
+	Max int `json:"max"`
+	// Link to the highlighted lines
+	Link string `json:"link,omitempty"`
+	// Pinned if the highlight changed
+	Pinned bool `json:"pinned,omitempty"`
+	// Error describing the problem.
+	Error string `json:"error,omitempty"`
+}
+
+var (
+	errNoHighlighter = errors.New("buildlog.highlighter unconfigured")
+)
+
+func analyzeArtifact(artifact api.Artifact, conf *parsedConfig) (*highlightResponse, error) {
+	if conf.highlighter == nil {
+		return nil, errNoHighlighter
+	}
+	link := artifact.CanonicalLink()
+	if !canSave(link) {
+		return nil, fmt.Errorf("Unsupported artifact: %q", link)
+	}
+	u, err := url.Parse(link)
+	if err != nil {
+		return nil, fmt.Errorf("parse artifact link %q: %v", link, err)
+	}
+	log := logrus.WithFields(logrus.Fields{
+		"artifact": link,
+	})
+
+	req := highlightRequest{
+		URL:       u.String(),
+		Pin:       conf.highlighter.Pin || conf.highlighter.Auto,
+		Overwrite: conf.highlighter.Overwrite,
+	}
+
+	buf, err := json.Marshal(req)
+	if err != nil {
+		log.WithError(err).Error("Failed to marshal highlight request")
```

Comment:
nit: what are the chances the error we send back to the caller also gets logged?

Reply:
Ack

There's basically zero chance this json.Marshal() fails, so not worth deduplicating.